python/pyarrow/scalar.pxi

```diff
@@ -1064,13 +1129,24 @@ cdef class MapScalar(ListScalar):
 
     def __getitem__(self, i):
         """
-        Return the value at the given index.
+        Return the value at the given index or key.
         """
+
         arr = self.values
         if arr is None:
-            raise IndexError(i)
-        dct = arr[_normalize_index(i, len(arr))]
-        return (dct[self.type.key_field.name], dct[self.type.item_field.name])
+            raise IndexError(i) if isinstance(i, int) else KeyError(i)
+
+        key_field = self.type.key_field.name
+        item_field = self.type.item_field.name
+
+        if isinstance(i, int):
+            dct = arr[_normalize_index(i, len(arr))]
+            return (dct[key_field], dct[item_field])
+        else:
+            for dct in arr:
+                if dct[key_field].as_py() == i:
+                    return dct[item_field]
+            raise KeyError(i)
```

Comment:
This is what I would suggest:

```suggestion
        if isinstance(i, (str, bytes)):
            i = self.keys().index(i)

        dct = arr[_normalize_index(i, len(arr))]
        return (dct[key_field], dct[item_field])
```

There will be two changes in behaviour though. The key lookup will return a key and value tuple and not only a value which is inline with the index lookup and I think it is valid. The error in case the key is not present (like "fake_key") will be raised by `index(i)` but with `ValueError`. That can be changed though.

Reply:
Done!